### PR TITLE
IEP-837: Adding the IDF env vars to the launch configs

### PR DESCRIPTION
**Pull request by alirana01**
## Description

Please include a summary of the change and which issue is fixed.

Fixes # ([IEP-837](https://jira.espressif.com:8443/browse/IEP-837))

## Type of change

Please delete options that are not relevant.
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How has this been tested?
Edit the launch configuration you should now see the env variables tags in the fields.
Try to run the launch configuration the launch and build should be working fine.
Change the IDF version and now try to run and build the project it should also be working fine.

Edit the launch configuration and change the available IDF_PATH and IDF_PYTHON_ENV_PATH to point to some static directories close and save the config and edit it again see if the changes are still there also try to launch it given that the paths are correct the launch should work.

**Test Configuration**:
* ESP-IDF Version: All
* OS (Windows,Linux and macOS): All

## Checklist
- [ ] PR Self Reviewed
- [ ] Applied Code formatting
- [ ] Verified on all platforms - Windows,Linux and macOS

### CODE REVIEW

**Review by alirana01 — COMMENTED** [review]
Some basic and main code areas are modified proper review for the code is required here by others as well

**Comment by kolipakakondal** [comment]
Following error appeared during the flash

> /Users/kondal/.espressif/python_env/idf5.2_py3.8_env/bin/python: can't open file '{IDF_PATH}/tools/idf.py': [Errno 2] No such file or directory

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/IDFUtil.java [hunk 1]
```diff
@@ -667,4 +675,9 @@ public static String getCurrentTarget()
 		}
 		return new SDKConfigJsonReader(project).getValue("IDF_TARGET"); //$NON-NLS-1$
 	}
+	
+	public static String getParseableVarValue(String var)
+	{
+		return "{" + var + "}"; //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (kolipakakondal) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/util/IDFUtil.java:681]: We can append the $ sign to follow the standard convention with other Eclipse IDE variables

**Review by kolipakakondal — COMMENTED** [review]
LGTM. We can incorporate the feedback discussed in the teams chat

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 2]
```diff
@@ -531,6 +533,78 @@ public boolean isValid(ILaunchConfiguration launchConfig) {
...
+		String pythonEnvPath = IDFUtil.getIDFPythonEnvPath();
+
+		if (expression.contains(idfPythonEnvReplaceableVarString) || expression.contains(idfPathReplaceableVarString)) {
+			expression = expression.replace(idfPathReplaceableVarString, idfPath);
+			expression = expression.replace(idfPythonEnvReplaceableVarString, pythonEnvPath);
+			return expression;
+		}
+		IStringVariableManager manager = VariablesPlugin.getDefault().getStringVariableManager();
+		return manager.performStringSubstitution(expression);
+	}
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:606]: we are copying a big chunk of code from the parent class just to introduce this method? I don't really understand why we need it. I think there is a mistake in DynamicVariableResolver led to this.
1. As I mentioned before, in SerifalFlashLaunchConfigDelegate it seems like we are not resolving the location field:
`String location = varManager.performStringSubstitution(IDFUtil.getIDFPythonEnvPath());` - by doing this, we are not using location attr from config.
2. If we expect a Python path in location, why not return it directly in DynamicVariableResolver, or introduce a new dynamicVariable, which will be more representative? For example, IDF_ENV_PYTHON or something like that

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 3]
```diff
@@ -550,7 +624,9 @@ public void performApply(ILaunchConfigurationWorkingCopy configuration) {
 
 			wc.setAttribute(IDFLaunchConstants.ATTR_JTAG_FLASH_ARGUMENTS, jtagArgumentsField.getText());
 			wc.setAttribute(IDFLaunchConstants.ATTR_SERIAL_FLASH_ARGUMENTS, uartAgrumentsField.getText());
+			wc.setAttribute(IDFLaunchConstants.ATTR_SERIAL_FLASH_ARGUMENTS_VIEW, uartAgrumentsField.getText());
 			wc.setAttribute(IDFLaunchConstants.ATTR_DFU_FLASH_ARGUMENTS, dfuArgumentsField.getText());
+			wc.setAttribute(IDFLaunchConstants.ATTR_DFU_FLASH_ARGUMENTS_VIEW, dfuArgumentsField.getText());
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:629]: I think should avoid these *VIEW fields. It complicates maintenance. I think it was a workaround in the first version of the PR that can be removed with a variables resolver

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 4]
```diff
@@ -812,15 +888,7 @@ private String getSerialPort() {
...
-		String location = IDFUtil.getIDFPythonEnvPath();
-		if (StringUtil.isEmpty(location)) {
-			try {
-				location = configuration.getAttribute(ICDTLaunchConfigurationConstants.ATTR_LOCATION,
-						SerialFlashLaunchConfigDelegate.getSystemPythonPath());
-			} catch (CoreException e) {
-				Logger.log(e);
-			}
-		}
+		String location = IDFUtil.getParseableVarValue(IDFEnvironmentVariables.IDF_PYTHON_ENV_PATH);
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:891]: newVariableExpression from the parent class can do this

**Review by sigmaaa — CHANGES_REQUESTED** [review]
Hi, @alirana01, thanks for the changes and for adding the dynamic variables resolver here. I think some areas can be still improved. Please, take a look on comments below

**Review by sigmaaa — CHANGES_REQUESTED** [review]
Hi, @alirana01, thanks for the changes and for adding the dynamic variables resolver here. I think some areas can be still improved. Please, take a look on comments below

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.espressif.idf.core;
+
+/**
+ * Class to contain all the variables that can be resolved by the IDFDynamicVariableResolver.java
+ * @author Ali Azam Rana
+ *
+ */
+public interface IDFDynamicVariables
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java:8]: in my opinion, using an enum here can increase readability. For example:
public enum IDFDynamicVariables {
IDF_PY,
IDF_PATH,
IDF_PYTHON_ENV_PATH
}

looks much cleaner 
But it is more personal preference, not necessary to change

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+{
+	/**
+	 * idf.py resolver variable
+	 */
+	String IDF_PY = "IDF_PY"; //$NON-NLS-1$
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java:13]: should be final

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	
+	/**
+	 * idf path variable added here to avoid redundancy in case of change in name
+	 */
+	String IDF_PATH = IDFEnvironmentVariables.IDF_PATH;
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java:18]: should be final

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+	
+	/**
+	 * idf virtual python executable path variable based on the env
+	 */
+	String IDF_PYTHON_ENV_PATH = IDFEnvironmentVariables.IDF_PYTHON_ENV_PATH;
```
Comment (sigmaaa) [inline bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java:23]: should be final

File: bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java (hunk 6 above)
Comment (alirana01) [reply bundles/com.espressif.idf.core/src/com/espressif/idf/core/IDFDynamicVariables.java:13]: I will change it to enum but variables in interfaces are public static and final by default.

**Comment by alirana01** [comment]
@sigmaaa Can you please review again? I have incorporated the changes you mentioned where possible

**Review by sigmaaa — COMMENTED** [review]
Hi, @alirana01, thanks for the changes. LGTM. It seems like we can remove validateLocation method from CMakeMainTab2 class since this method from the parent class does the same thing. Also, IMHO we don't need a new method in IDFUtil class.

In the comments, I mentioned a few issues, that don't belong to this PR, so we can handle these problems here, or create new Jira tickets for them. I will summarize my findings here: 
 1. Location field can't be changed from the UI
 2. In the SerialFlashLaunchConfigDelegate class in launchInternal method, we do not rely on the location field in the first place, and relying on IDFUtil.getIDFPythonEnvPath(); instead, which is wrong in my opinion
3. Instead of handling string substitution for openocd, serial port directly in  SerialFlashLaunchConfigDelegate, we can introduce new dynamic variables and resolver class for this goal

File: bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java [hunk 9]
```diff
@@ -618,8 +680,8 @@ private void updateArgumentsWithDefaultFlashCommand(ILaunchConfiguration configu
 		try {
 			String uartFlashCommand = configuration.getAttribute(IDFLaunchConstants.ATTR_SERIAL_FLASH_ARGUMENTS,
 					StringUtil.EMPTY);
-			uartAgrumentsField
-					.setText(uartFlashCommand.isBlank() ? ESPFlashUtil.getEspFlashCommand(ESPFlashUtil.SERIAL_PORT)
+			uartAgrumentsField.setText(
+					uartFlashCommand.isBlank() ? ESPFlashUtil.getParseableEspFlashCommand(ESPFlashUtil.SERIAL_PORT)
```
[the remark this answers is not in the document]
Comment (sigmaaa) [reply bundles/com.espressif.idf.launch.serial.ui/src/com/espressif/idf/launch/serial/ui/internal/CMakeMainTab2.java:684]: nevermind, I missed the place with a comment. I meant, that we can use newVariableExpression in CMakeMainTab instead IDFUtil.resolveExpressionFromVariableManager.

**Review by sigmaaa — APPROVED** [review]
LGTM

**Comment by sigmaaa** [comment]
Hi @AndriiFilippov, please take a look. Also as test cases, you try to flash via different interfaces (UART, JTAG, DFU) since the commands for them are affected by variable

**Comment by AndriiFilippov** [comment]
@alirana01 hi! 
Tested under: 
OS - Windows 10 

Once I open IDE - get error :


image



my Java:
image

**Comment by AndriiFilippov** [comment]
@alirana01 hi ! 

Tested under: 
OS - Windows 10 / Linux 
esp-idf: master 

image

able to see the env variables tags in the fields.
run the launch configuration the launch and build working fine 👍 
change the IDF version -> run and build the project -> working fine 👍 
able to flash via UART & JTAG 👍 

But I was not able to change the available IDF_PATH and IDF_PYTHON_ENV_PATH to point to some static directories. Once I change it ->save and close Launch Config -> reopen Launch Config -> Python Path delete my changes 

image

**Comment by alirana01** [comment]
> @alirana01 hi !
> 
> Tested under: OS - Windows 10 / Linux esp-idf: master
> But I was not able to change the available IDF_PATH and IDF_PYTHON_ENV_PATH to point to some static directories. Once I change it ->save and close Launch Config -> reopen Launch Config -> Python Path delete my changes
> 
> image

@AndriiFilippov you can try with the latest build I have pushed the required changes

**Comment by alirana01** [comment]
@AndriiFilippov I have also synced the branch with master so hopefully after most recent builds you wont see the error when starting the IDE

**Comment by kolipakakondal** [comment]
HI @alirana01 Can you rebase with master to avoid conflicts?

**Comment by AndriiFilippov** [comment]
@alirana01 @sigmaaa hi! 
Tested under: 
OS - Windows10

Steps to reproduce: open Launch Configuration -> insert wrong path to Location field -> press "Duplicate" -> open duplicate configuration - it saves the wrong python path - still able to build with wrong path :) 

image

**Pull request merged**